keylime/da/attest.py

```diff
@@ -0,0 +1,154 @@
+        end_date = rmc.end_of_times
+    else:
+        end_date = int(datetime.fromisoformat(args.end_date).strftime("%s"))
+
+    if args.start_date == "first":
+        start_date = rmc.start_of_times
+    elif args.start_date == "none":
+        start_date = end_date - 1
+    else:
+        start_date = int(datetime.fromisoformat(args.start_date).strftime("%s"))
+
+    if end_date != rmc.end_of_times or start_date != rmc.end_of_times - 1:
+        logger.info(
+            "=> Selected date range for attestation records is: start %d (%s) and end %d (%s)",
+            start_date,
+            datetime.fromtimestamp(start_date).strftime("%Y-%m-%dT%H:%M:%S.%f%z"),
+            end_date,
+            datetime.fromtimestamp(end_date).strftime("%Y-%m-%dT%H:%M:%S.%f%z"),
+        )
+    else:
+        logger.info("=> Will only extract the very last attestation record found on the persistent store")
+
+    if args.agent_uuid.lower() == "all":
+        logger.info("=> Getting a list of agents registered on the persistent store")
+        agent_list = rmc.agent_list_retrieval()
+    else:
+        logger.info("=> Focusing on the agent %s on the persistent store.", args.agent_uuid)
+        agent_list = [args.agent_uuid]
+
+    for agent_uuid in agent_list:
+        logger.info("===> Getting all existing registration records for agent %s...", agent_uuid)
+
+        ak_list = rmc.build_key_list(agent_uuid)
+
+        if not ak_list:
+            logger.error("Unable to assemble a list of AKs for the agent agent %s", agent_uuid)
+            return
+
+        logger.info("===> Getting all existing attestation records for agent %s ...", agent_uuid)
+        attestation_record_list = rmc.record_read(agent_uuid, start_date, end_date, "verifier")
+
+        logger.info("=====> Verifing the state of agent %s over time...", agent_uuid)
+
+        p_tpm_ts = 0
+        c_tpm_ts = 0
+        d_tpm_ts = 0
+
+        agentAttestState = None
+        for attestation_record in attestation_record_list:
+            agent = attestation_record["agent"]
+            json_response = attestation_record["json_response"]
+            ima_policy = attestation_record["ima_policy"]
+
+            logger.info(
+                "----------- Attesting data (quote and logs from %s, captured by verifier %s (%s:%s)",
+                agent_uuid,
+                agent["verifier_id"],
+                agent["verifier_ip"],
+                agent["verifier_port"],
+            )
+
+            if not agentAttestState:
+                agentAttestState = cloud_verifier_common.get_AgentAttestStates().get_by_agent_id(agent["agent_id"])
+
+            if "tpm_clockinfo" in agent:
+                if "clock" in agent["tpm_clockinfo"]:
+                    p_tpm_ts = agent["tpm_clockinfo"]["clock"]
+
+            failure = cloud_verifier_common.process_quote_response(
+                agent, ima_policy, json_response["results"], agentAttestState
+            )
```

Comment:
Since I am not very familiar with DA: So you are replaying the measurements here to presumably get the same result as when it was done live. Does this also deal with a machine that was rebooted and attestation was started over? What happens if we now were to support updating a policy -- does it know how to deal with that also?

Reply:
My guess is that it would be able to deal with an updated policy following line 99 -- though each record must be huge then if we have an IMA policy with thousands of attestation entries that is stored alongside every measurement...

Reply:
On this PR, we store the actual IMA policy within the "historical" record (required for replay) and thus they can be "large" yes. In a subsequent PR, I will add the ability to **replace** the stored IMA policy with an updated one.

Reply:
Another complication may arise if someone restarts attestation - how do you recognize that?